File: 08_02_Robot_in_a_Grid.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int>>& obstacleGrid) {        
        // grid[i][j] =  obstacle ? grid[i - 1][j] + grid[i][j - 1] : 0;
        auto n = obstacleGrid.size();
        auto m = obstacleGrid[0].size();
        vector<vector<int>> grid(n,vector<int>(m));
        grid[0][0] = 1;
        for (auto i = 0; i < n; i++) {
            for (auto j = 0; j < m; j++) {        
                if (obstacleGrid[i][j] == 1) {
                    grid[i][j] = 0;
                    continue;
                }

                if (i > 0) {
                    grid[i][j] += grid[i - 1][j];
                }
 
                if (j > 0) {
                    grid[i][j] += grid[i][j - 1];         
                }
            }
        }
        return grid[n - 1][m - 1];
    }
};

File: 08_02_Robot_in_a_Grid.java
class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        int[][] dp = new int[obstacleGrid.length][obstacleGrid[0].length];
        dp[0][0] = 1;
        
        
        for (int i = 0;i < obstacleGrid.length;i++) {
            for (int j = 0;j < obstacleGrid[0].length;j++) {
                if (obstacleGrid[i][j] == 1) {
                    dp[i][j] = 0;
                    continue;
                }
                if (i != 0 && obstacleGrid[i - 1][j] != 1) {
                    dp[i][j] += dp[i - 1][j];
                }
                
                if (j != 0 && obstacleGrid[i][j - 1] != 1) {
                    dp[i][j] += dp[i][j - 1];
                }
            }
        }
        
        return dp[dp.length - 1][dp[0].length - 1];
    }
}
